### PR TITLE
home-assistant: handle packages with extra requirements

### PR DESCRIPTION
**Pull request by mweinelt**
###### Description of changes

https://github.com/NixOS/nixpkgs/pull/167560#issuecomment-1090978534

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/samsungctl/default.nix [hunk 1]
```diff
@@ -11,12 +15,17 @@ buildPythonPackage rec {
...
+  passthru.extras-require = {
+    websocket = [
+      websocket-client
+    ];
+    # interactive_ui requires curses package
+  };
 
   # no tests
-  doCheck = false;
+  doCheck = true;
```
Comment (dotlambda) [inline pkgs/development/python-modules/samsungctl/default.nix:26]: `checkPhase` should be specified or `pytestCheckHook` added to `checkInputs`
no need to specify `doCheck = true`

File: pkgs/development/python-modules/androidtv/default.nix [hunk 2]
```diff
@@ -25,14 +25,24 @@ buildPythonPackage rec {
...
     pure-python-adb
   ];
 
+  passthru.extras-require = {
+    async = [
+      aiofiles
+    ];
+    usb = [
+      adb-shell.extras-require.usb
+    ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/androidtv/default.nix:37]:
```suggestion
    inherit (adb-shell.extras-require) usb;
```

File: pkgs/servers/home-assistant/component-packages.nix [hunk 3]
```diff
@@ -94,7 +94,7 @@
     "bloomsky" = ps: with ps; [ ];
     "blueprint" = ps: with ps; [ ];
     "bluesound" = ps: with ps; [ xmltodict ];
-    "bluetooth_le_tracker" = ps: with ps; [ pygatt ];
+    "bluetooth_le_tracker" = ps: with ps; [ pygatt pygatt.extras-require.GATTTOOL ];
```
Comment (dotlambda) [inline pkgs/servers/home-assistant/component-packages.nix:97]:
```suggestion
    "bluetooth_le_tracker" = ps: with ps; [ pygatt ] ++ pygatt.extras-require.GATTTOOL;
```
would be nice but I guess this is fine.

Comment (mweinelt) [reply pkgs/servers/home-assistant/component-packages.nix:97]: Oh yeah, they're nested lists now. Shoot.

File: pkgs/development/python-modules/samsungctl/default.nix (hunk 1 above)
Comment (mweinelt) [reply pkgs/development/python-modules/samsungctl/default.nix:26]: Yeah, just leftovers from when I was checking out the tests.

Comment (dotlambda) [reply pkgs/development/python-modules/samsungctl/default.nix]: Shouldn't the dependencies be in propagatedBuildInputs if they're required for importing the module?

Comment (mweinelt) [reply pkgs/development/python-modules/samsungctl/default.nix]: Spot on.

File: pkgs/servers/home-assistant/component-packages.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/servers/home-assistant/component-packages.nix]: Okay, I kinda went for a full reformat. PTAL.

**Comment by mweinelt** [comment]
Result of `nixpkgs-review pr 167630` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>ghost</li>
    <li>home-assistant-component-tests.androidtv</li>
    <li>home-assistant-component-tests.bluetooth_le_tracker</li>
    <li>home-assistant-component-tests.samsungtv</li>
    <li>home-assistant-component-tests.tradfri</li>
    <li>python310Packages.adb-shell</li>
    <li>python310Packages.androidtv</li>
    <li>python310Packages.meshtastic</li>
    <li>python310Packages.pure-python-adb</li>
    <li>python310Packages.py3rijndael</li>
    <li>python310Packages.pygatt</li>
    <li>python310Packages.pytradfri</li>
    <li>python310Packages.samsungctl</li>
    <li>python310Packages.samsungtvws</li>
    <li>python39Packages.adb-shell</li>
    <li>python39Packages.androidtv</li>
    <li>python39Packages.meshtastic</li>
    <li>python39Packages.pure-python-adb</li>
    <li>python39Packages.py3rijndael</li>
    <li>python39Packages.pygatt</li>
    <li>python39Packages.pytradfri</li>
    <li>python39Packages.samsungctl</li>
    <li>python39Packages.samsungtvws</li>
  </ul>
</details>

File: pkgs/servers/home-assistant/component-packages.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/servers/home-assistant/component-packages.nix]: Merging in a bit so I can continue updating home-assistant (two minor releases) based on these changes.

File: pkgs/development/python-modules/pure-python-adb/default.nix [hunk 4]
```diff
@@ -11,11 +17,23 @@ buildPythonPackage rec {
...
+
+  checkInputs = [
+    pytestCheckHook
+  ]
+  ++ passthru.extras-require.async;
+
+  pythonImportsCheck = lib.optionals doCheck [
+    "ppadb.client"
+    "ppadb.client_async"
+  ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/pure-python-adb/default.nix:36]:
```suggestion
  pythonImportsCheck = [
    "ppadb.client"
  ] ++ lib.optionals doCheck [
    "ppadb.client_async"
  ];
```

Comment (mweinelt) [reply pkgs/development/python-modules/pure-python-adb/default.nix]: Again … good catch! Thank you.

**Pull request merged**